pkg/certificate/providers/mrc.go

```diff
@@ -0,0 +1,127 @@
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/k8s/informers"
+)
+
+var (
+	errNilInformerCollection = errors.New("informer collection is nil")
+)
+
+var mrcKeyFunc = func(mrc v1alpha2.MeshRootCertificate) string {
+	return fmt.Sprintf("%s/%s", mrc.GetNamespace(), mrc.GetName())
+}
+
+// MRCComposer is a composer object that allows consumers
+// to observe MRCs (via List() and Watch()) as well as generate
+// `certificate.Provider`s from those MRCs
+type MRCComposer struct {
+	informerCollection *informers.InformerCollection
+	MRCProviderGenerator
+}
+
+// List returns the MRCs stored in the informerCollection's store
+func (m *MRCComposer) List() ([]*v1alpha2.MeshRootCertificate, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+	// informers return slice of pointers so we'll convert them to value types before returning
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	var mrcs []*v1alpha2.MeshRootCertificate
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcs = append(mrcs, mrc)
+	}
+
+	return mrcs, nil
+}
+
+// Watch returns a channel that receives events whenever MRCs are added, updated, and deleted
+// from the informerCollection's MRC store. Channels returned from multiple invocations of
+// Watch() are unique and have no coordination with each other. Events are guaranteed
+// to be ordered for any particular resources, but NOT across different resources.
+// Watch() should only send new events as of the time the Watch call returns
+func (m *MRCComposer) Watch(ctx context.Context) (<-chan certificate.MRCEvent, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+
+	// To make sure we only send new events, we cache the object versions
+	// before the eventhandler is added
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	mrcObjectVersionCache := map[string]int64{}
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcObjectVersionCache[mrcKeyFunc(*mrc)] = mrc.GetGeneration()
+	}
+
+	eventChan := make(chan certificate.MRCEvent)
+	m.informerCollection.AddEventHandler(informers.InformerKeyMeshRootCertificate, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			mrc := obj.(*v1alpha2.MeshRootCertificate)
+			if !checkGeneration(mrcObjectVersionCache, *mrc) {
+				return
+			}
+			eventChan <- certificate.MRCEvent{
+				Type: certificate.MRCEventAdded,
+				MRC:  mrc,
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldMRC := oldObj.(*v1alpha2.MeshRootCertificate)
+			mrc := newObj.(*v1alpha2.MeshRootCertificate)
+			if !checkGeneration(mrcObjectVersionCache, *mrc) {
```

Comment:
(If we find we still need to do this)

Would an update or delete event ever come through that happened before the event handler is added? My understanding is that every existing object comes through as an "add" event to the handler, then any update or delete is based off of those objects which would have to happen after the handler was created.